### ISSUE TITLE
Synchronous Expiry Listener

### ISSUE THREAD

**Issue by cruftex** [cache2k/cache2k#127]
In cache2k version 1.2 expiry listeners are always asynchronous. For version 1.4 support for synchronous expiry listeners is planned.

As it turns out, for supporting synchronous expiry some more restructuring and careful testing is needed.

**Comment by cruftex** [comment]
I am not 100% sure what the guaranteed semantics of a synchronous expiry should be.

Since synchronous, nothing else can happen with the entry until the listener is completed, which means subsequent operations must block.

In cache2k expiry might be "sharp" which means, expiry happens when a point in time is reached.
Now there are different possibilities, when the event must be delivered:

1. When expired. That's not possible, since time always flows
2. Before detected. For example, `cache.contains()` returns  false.
3. Before another mutation is done or the loader is about to start.
4. The expiry listener isn't called when mutation or loader is started and the event thread which would send the expiry event makes no progress before. That means, an expiry can happen silently.

@ben-manes any opinion?

**Comment by cruftex** [comment]
@ben-manes ohh, from your answer I feel it might get strenuous to sort this one out. Well better then banging the head just on my own.

> However a synchronous writer is incompatible with an asynchronous cache

Now we have the word synchronous with two different meanings :(
I would expect a writer always being synchronous with respect to the operation, meaning that the operation isn't finished successfully when the writer is not finished yet. Whether the writer has a synchronous or asynchronous execution model is a different aspect.

In cache2k, I basically took over the idea of synchronous and asynchronous events from JSR107/JCache and it seemed/seems to make sense. The JCache API says for `CacheEntryListenerConfiguration.isSynchronous()`

> Determines if the thread that caused an event to be created should be
> blocked (not return from the operation causing the event) until 
> `CacheEntryListener` has been notified.

Things are straight forward for insert, update and remove. Expiry is different since racy and there are different expectations or "underspecified" APIs. BTW, I vaguely followed the issues in Caffeine about this.

Why I am bothering? Let's do an example with Caffeine code:

```java
final AtomicBoolean removed = new AtomicBoolean(false);
Cache c =
  Caffeine.newBuilder()
    .maximumSize(1000)
    .expireAfterWrite(2, TimeUnit.SECONDS)
    .writer(new CacheWriter<Object, Object>() {
      @Override
      public void write(@NonNull final Object key, @NonNull final Object value) {

      }

      @Override
      public void delete(@NonNull final Object key, @Nullable final Object value, @NonNull final RemovalCause cause) {
        removed.set(true);
      }
    })
    .build();
long t0 = System.currentTimeMillis();
c.put(1,2);
while (c.asMap().containsKey(1)) { }
System.out.println("Time elapsed: " + (System.currentTimeMillis() - t0) + ", removed=" + removed.get());
```

Results in:

```
Time elapsed: 2000, removed=false
```

Which is a violation of the second strongest guarantee I listed above. The `RemovalListener` is doing the same. In case the event is used for resource cleanup, new resources could be requested after the previous ones are cleaned via the event. BTW: I didn't do this test before I did state the question, maybe I should have done that before.

I think, it's "desirable" that the event is executed before the expiry becomes visible. However, that means `containsKey()` needs to block until the listener/writer is completed. That's a "no no". Another approach could be, in case there are listeners/writers that the actual expiry is hold until event completion.

**Comment by ben-manes** [comment]
> I would expect a writer always being synchronous with respect to the operation, meaning that the operation isn't finished successfully when the writer is not finished yet. Whether the writer has a synchronous or asynchronous execution model is a different aspect.

In my case the writer is part of the map's atomic computation. A `put(key, value)` would transform into a `compute` that invokes the `CacheWriter#write` and, if successful, returns the new value. Similarly a removal, e.g. by eviction, is a `computeIfPresent` that invokes the `CacheWriter#delete` when discarding the mapping.

In an asynchronous cache we store a `CompletableFuture<V>` as the value. The map computation for the `put` case is not very interesting to intercept if the future is in-flight. An expensive writer that blocks per-entry map operations is likely the antithesis of the user's goal if they want to minimize blocking calls that go through the cache.

The primary use-case for our `CacheWriter` is blocking calls for synchronous integration between different things, like an auxiliary data structure. So it expands the size of the map's atomic block, whereas the asynchronous cache is meant to minimize the atomic block's size.

> BTW, I vaguely followed the issues in Caffeine about this.

That was really about how prompt the notification is when the entry is removed. By default when an entry expires, no thread is scheduled to remove it and call a listener. Instead the entry can no longer be retrieved and treated as softly removed (visible by `Map.size` but treated as absent when performing operations like `Map.put`). The entry will be removed and the listener notified based on other cache activity, which triggers a routine maintenance cycle.

The user instead wanted a thread to be scheduled based on the expiration time, the entry removed, and a notification sent. That way if the cache is idle the expiration work is handled promptly. Therefore a new `Scheduler` setting is available to provide that scheduling thread, with an option to utilize Java 9's global one if possible.

> In case the event is used for resource cleanup, new resources could be requested after the previous ones are cleaned via the event.

The map entry was not removed, but hidden. A subsequent write to the cache for that or another mapping, like `cache.put(1, 3)`, will cause the entry to be removed. The `RemovalListener` would be notified after the mapping was established, so `1=>3` is visible. The `CacheWriter` would be notified during the `Map.put` so that `1=>3` waits for the writer to finish before the new mapping is established.

> I think, it's "desirable" that the event is executed before the expiry becomes visible. However, that means containsKey() needs to block until the listener/writer is completed.

I separate out my thinking in terms of the mapping versus the notification. The entry is no longer visible once it has expired, but whether its actually removed and notification propagated is separate.

A cache is a probabilistic data structure and multiple threads accessing it should lead to an expectation of some non-determinism. I don't think that blocking `containsKey()` to perform a possibly expensive write/listener operation is generally desirable. It starts getting into strict consistency expectations which I don't think is reasonable. I think per-entry linearizability of writes is a good model to try to offer, though.

**Comment by cruftex** [comment]
I agree on your conclusion. We should send the expiry event before a mutation. So whenever the user is doing something which is supposed to be atomic on that entry, e.g. a compute or load, the expiry event should be sent. Especially, in a read through configuration, it might be surprising that a reload is done because of an expiry, but the expiry event is not received before. We should add this to the documentation somewhere.

**Comment by cruftex** [comment]
Completed and shipped in 1.3.1.Alpha. Had undergone extensive testing and updates can be found in 1.3.2.Aplha
